nameresolution/consul/consul.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	return resolverCfg, nil
+}
+
+func getClientConfig(cfg configSpec) *consul.Config {
+	// If no client config use library defaults
+	if cfg.ClientConfig != nil {
+		return cfg.ClientConfig
+	}
+
+	return consul.DefaultConfig()
+}
+
+func getRegistrationConfig(cfg configSpec, props map[string]string) (*consul.AgentServiceRegistration, error) {
+	// if advanced registration configured ignore other registration related configs
+	if cfg.AdvancedRegistration != nil {
+		return cfg.AdvancedRegistration, nil
+	} else if !cfg.SelfRegister {
+		return nil, nil
+	}
+
+	var appID string
+	var appPort string
+	var host string
+	var httpPort string
+	var ok bool
+
+	if appID, ok = props[nr.AppID]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppID)
+	}
+
+	if appPort, ok = props[nr.AppPort]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppPort)
+	}
+
+	if host, ok = props[nr.HostAddress]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.HostAddress)
+	}
+
+	if httpPort, ok = props[nr.DaprHTTPPort]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.DaprHTTPPort)
+	}
+
+	// if no health checks configured add dapr sidecar health check by default
+	if len(cfg.Checks) == 0 {
+		cfg.Checks = []*consul.AgentServiceCheck{
+			{
+				Name:     "Dapr Health Status",
+				CheckID:  fmt.Sprintf("daprHealth:%s", appID),
+				Interval: "15s",
+				HTTP:     fmt.Sprintf("http://%s:%s/v1.0/healthz", host, httpPort), // default assumes consul agent on svc host
+			},
+		}
+	}
+
+	appPortInt, err := strconv.Atoi(appPort)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing app port: %w", err)
+	}
+
+	return &consul.AgentServiceRegistration{
+		Name:    appID,
+		Address: host,
+		Port:    appPortInt,
+		Checks:  cfg.Checks,
+		Tags:    cfg.Tags,
+		Meta:    cfg.Meta,
+	}, nil
+}
+
+func getQueryOptionsConfig(cfg configSpec) *consul.QueryOptions {
+	// if no query options configured add default filter matching every tag in config
+	if cfg.QueryOptions == nil {
+		return &consul.QueryOptions{
+			UseCache: true,
+		}
+	}
+
+	return cfg.QueryOptions
+}
+
+func parseConfig(rawConfig interface{}) (configSpec, error) {
+	config := configSpec{}
+	rawConfig = convertGenericConfig(rawConfig)
+
+	data, err := json.Marshal(rawConfig)
+	if err != nil {
+		return config, fmt.Errorf("error serializing to json: %w", err)
+	}
+
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.DisallowUnknownFields()
+
+	if err := decoder.Decode(&config); err != nil {
+		return config, fmt.Errorf("error deserializing to configSpec: %w", err)
+	}
+
+	return config, nil
+}
+
+// helper function for transforming interface{} into json serializable
+func convertGenericConfig(i interface{}) interface{} {
+	switch x := i.(type) {
+	case map[interface{}]interface{}:
+		m2 := map[string]interface{}{}
+		for k, v := range x {
+			m2[k.(string)] = convertGenericConfig(v)
```

Comment:
In the unlikely scenario that this is not a `string` it will panic.

Reply:
done